File: app/src/main/java/com/example/myapplication/data/local/LoggedFoodDao.kt
package com.example.myapplication.data.local

import androidx.room.*
import kotlinx.coroutines.flow.Flow

@Dao
interface LoggedFoodDao {
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(loggedFood: LoggedFoodEntity): Long

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(loggedFoods: List<LoggedFoodEntity>)

    @Query("SELECT * FROM logged_foods WHERE id = :id LIMIT 1")
    suspend fun getById(id: Long): LoggedFoodEntity?

    @Query("DELETE FROM logged_foods WHERE id = :id")
    suspend fun delete(id: Long): Int

    @Query("SELECT * FROM logged_foods WHERE epochDay = :epochDay ORDER BY timestamp ASC")
    fun observeDay(epochDay: Long): Flow<List<LoggedFoodEntity>>

    @Query("SELECT * FROM logged_foods WHERE epochDay = :epochDay ORDER BY timestamp ASC")
    suspend fun dayNow(epochDay: Long): List<LoggedFoodEntity>

    @Query("SELECT * FROM logged_foods ORDER BY timestamp DESC LIMIT :limit")
    fun observeRecentFoods(limit: Int): Flow<List<LoggedFoodEntity>>

    @Query("SELECT * FROM logged_foods ORDER BY timestamp DESC LIMIT :limit")
    suspend fun recentFoodsNow(limit: Int): List<LoggedFoodEntity>

    @Query("DELETE FROM logged_foods")
    suspend fun deleteAll()
}

File: flutter/lib/data/local/daos/logged_food_dao.dart
import 'package:drift/drift.dart';
import '../database.dart';
import '../tables/logged_foods.dart';

part 'logged_food_dao.g.dart';

@DriftAccessor(tables: [LoggedFoods])
class LoggedFoodDao extends DatabaseAccessor<GymDatabase>
    with _$LoggedFoodDaoMixin {
  LoggedFoodDao(super.db);

  Future<int> insert(LoggedFoodsCompanion loggedFood) {
    return into(loggedFoods)
        .insert(loggedFood, mode: InsertMode.insertOrReplace);
  }

  Future<void> insertAll(List<LoggedFoodsCompanion> loggedFoodsList) async {
    await batch((b) {
      for (final f in loggedFoodsList) {
        b.insert(loggedFoods, f, mode: InsertMode.insertOrReplace);
      }
    });
  }

  Future<LoggedFoodData?> getById(int id) {
    final query = select(loggedFoods)
      ..where((tbl) => tbl.id.equals(id))
      ..limit(1);
    return query.getSingleOrNull();
  }

  Future<int> deleteLoggedFood(int id) {
    final query = delete(loggedFoods)..where((tbl) => tbl.id.equals(id));
    return query.go();
  }

  Stream<List<LoggedFoodData>> observeDay(int epochDay) {
    final query = select(loggedFoods)
      ..where((tbl) => tbl.epochDay.equals(epochDay))
      ..orderBy([(tbl) => OrderingTerm.asc(tbl.timestamp)]);
    return query.watch();
  }

  Future<List<LoggedFoodData>> dayNow(int epochDay) {
    final query = select(loggedFoods)
      ..where((tbl) => tbl.epochDay.equals(epochDay))
      ..orderBy([(tbl) => OrderingTerm.asc(tbl.timestamp)]);
    return query.get();
  }

  Stream<List<LoggedFoodData>> observeRecentFoods(int limit) {
    final query = select(loggedFoods)
      ..orderBy([(tbl) => OrderingTerm.desc(tbl.timestamp)])
      ..limit(limit);
    return query.watch();
  }

  Future<List<LoggedFoodData>> recentFoodsNow(int limit) {
    final query = select(loggedFoods)
      ..orderBy([(tbl) => OrderingTerm.desc(tbl.timestamp)])
      ..limit(limit);
    return query.get();
  }

  Future<int> deleteAll() {
    return delete(loggedFoods).go();
  }
}
